### PR TITLE
test(reasoning): add comprehensive test coverage for ReasoningContext

### LINKED ISSUE #1833: [Test] Test Coverage for Reasoning Mechanism
## Description

Add unit tests and functional tests for reasoning mechanism.

Part of #1829

Depends on: #1834 (HGraph Reasoning Integration)

## Requirements

- [ ] Unit tests for ReasoningContext data recording
- [ ] Unit tests for ReasoningContext report generation
- [ ] Functional tests for found/missed analysis correctness
- [ ] Performance tests verifying zero overhead when disabled

### Unit Tests
- `ReasoningContextTest.RecordVisit` - Verify visit tracking
- `ReasoningContextTest.RecordEviction` - Verify eviction tracking
- `ReasoningContextTest.Diagnose` - Verify diagnosis logic
- `ReasoningContextTest.GenerateReport` - Verify JSON output

### Functional Tests
- Verify reasoning enabled when `expected_labels` non-empty
- Verify found labels appear in report with correct rank
- Verify missed labels appear with correct diagnosis
- Verify reasoning disabled doesn't affect search correctness

### Performance Tests
- Benchmark search with reasoning disabled vs baseline
- Confirm no measurable overhead

## Acceptance Criteria

- [ ] All unit tests pass
- [ ] Functional tests cover main use cases
- [ ] Performance test confirms zero overhead

## Related

- Parent issue: #1829
- Depends on: #1834 (HGraph Reasoning Integration)

### PR DESCRIPTION
**Pull request by LHT129**
## Description

Add comprehensive test coverage for the ReasoningContext reasoning mechanism.

### Unit tests (search_reasoning_test.cpp)
- SetTermination: verify storage and overwrite
- MarkResult: verify matching targets and non-expected ID handling
- SetTrueDistance: verify update and non-expected ID handling
- RecordVisit edge case: first distance preserved, hop updates on revisit
- RecordEviction edge case: earlier visit hop preserved
- GenerateReport found count: all-found scenario reports correctly

### Functional tests (test_hgraph.cpp)
- Found Verification: expected label in results shows correct found count
- Multiple Labels Mixed: mix of found/unfound generates valid report
- Results Not Affected: identical results with and without reasoning
- Empty Expected Labels: no reasoning analysis when expected_labels empty
- Zero Overhead When Disabled: performance parity (< 20% overhead)

Closes #1833

## Test Plan
- [x] All 12 reasoning unit tests pass (78 assertions)
- [x] All 5 new + 1 existing reasoning functional tests pass (154+11 assertions)
- [x] Zero overhead test confirms < 20% latency difference

### CODE REVIEW

File: src/impl/reasoning/search_reasoning_test.cpp [hunk 1]
```diff
@@ -271,4 +271,151 @@ TEST_CASE("ReasoningContext termination reasons are centralized", "[reasoning]")
...
+
+    ctx.InitializeExpectedTargets(labels, label_to_inner_id);
+
+    ctx.RecordVisit(0, 0.3F, 1);
+    ctx.RecordVisit(0, 0.5F, 2);
+
+    auto it = ctx.expected_traces_.find(0);
+    REQUIRE(it.value().quantized_distance == 0.3F);
+    REQUIRE(it.value().visited_at_hop == 2);
+}
```
[the remark this answers is not in the document]
Comment (LHT129) [reply src/impl/reasoning/search_reasoning_test.cpp:384]: Good catch. Added a SECTION for zero first distance. The test documents the current behavior: RecordVisit uses 0.0F as sentinel, so a true zero distance gets overwritten by the next visit. Fixed in f347b9d6.

File: tests/test_hgraph.cpp [hunk 2]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+        REQUIRE(r.has_value());
+    }
+    auto end_no_reasoning = std::chrono::high_resolution_clock::now();
+    auto no_reasoning_us =
+        std::chrono::duration_cast<std::chrono::microseconds>(end_no_reasoning - start_no_reasoning)
+            .count();
+
+    double ratio =
+        static_cast<double>(no_reasoning_us) / static_cast<double>(std::max(baseline_us, 1L));
+    REQUIRE(ratio < 1.2);
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Good point. Changed baseline to use KnnSearch API directly so we actually compare SearchWithRequest overhead against the standard search path. Fixed in f347b9d6.

File: tests/test_hgraph.cpp [hunk 3]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    req_baseline.query_ = query;
+
+    constexpr int warmup_rounds = 10;
+    constexpr int measure_rounds = 50;
+
+    for (int i = 0; i < warmup_rounds; ++i) {
+        index->SearchWithRequest(req_baseline);
+    }
+
+    auto start_baseline = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Switched to steady_clock. Fixed in 487bd676.

File: tests/test_hgraph.cpp [hunk 4]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    for (int i = 0; i < measure_rounds; ++i) {
+        auto r = index->SearchWithRequest(req_baseline);
+        REQUIRE(r.has_value());
+    }
+    auto end_baseline = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Switched to steady_clock. Fixed in 487bd676.

File: tests/test_hgraph.cpp [hunk 5]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    req_no_reasoning.topk_ = 10;
+    req_no_reasoning.params_str_ = fmt::format(fixtures::search_param_tmp, 100, false);
+    req_no_reasoning.query_ = query;
+    req_no_reasoning.expected_labels_ = {};
+
+    for (int i = 0; i < warmup_rounds; ++i) {
+        index->SearchWithRequest(req_no_reasoning);
+    }
+
+    auto start_no_reasoning = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Switched to steady_clock. Fixed in 487bd676.

File: tests/test_hgraph.cpp [hunk 6]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    for (int i = 0; i < measure_rounds; ++i) {
+        auto r = index->SearchWithRequest(req_no_reasoning);
+        REQUIRE(r.has_value());
+    }
+    auto end_no_reasoning = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Switched to steady_clock. Fixed in 487bd676.

File: tests/test_hgraph.cpp [hunk 7]
```diff
@@ -876,6 +877,216 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    auto baseline_us =
+        std::chrono::duration_cast<std::chrono::microseconds>(end_baseline - start_baseline)
+            .count();
+
+    vsag::SearchRequest req_no_reasoning;
+    req_no_reasoning.topk_ = 10;
+    req_no_reasoning.params_str_ = fmt::format(fixtures::search_param_tmp, 100, false);
+    req_no_reasoning.query_ = query;
+    req_no_reasoning.expected_labels_ = {};
+
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp:1067]: Fixed: baseline now uses KnnSearch API directly instead of SearchWithRequest. Fixed in 487bd676.

File: tests/test_hgraph.cpp [hunk 8]
```diff
@@ -876,6 +877,213 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+        REQUIRE(r.has_value());
+    }
+    auto end_no_reasoning = std::chrono::steady_clock::now();
+    auto no_reasoning_us =
+        std::chrono::duration_cast<std::chrono::microseconds>(end_no_reasoning - start_no_reasoning)
+            .count();
+
+    double ratio =
+        static_cast<double>(no_reasoning_us) / static_cast<double>(std::max(baseline_us, 1L));
+    REQUIRE(ratio < 1.2);
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp]: Good point. Increased warmup to 20, measurement rounds to 100, and relaxed threshold to 1.5x to reduce flakiness. Fixed in 34b0c4ea.

File: tests/test_hgraph.cpp [hunk 9]
```diff
@@ -876,6 +877,213 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+
+    auto baseline = index->SearchWithRequest(req);
+    REQUIRE(baseline.has_value());
+    REQUIRE(baseline.value()->GetDim() > 0);
+
+    auto* ids = baseline.value()->GetIds();
+    int64_t found_label = ids[0];
+    int64_t unlikely_label = -999;
+
+    req.expected_labels_ = {found_label, unlikely_label};
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp:947]: Agreed. Changed to derive the missing label from dataset size: `dataset->base_->GetNumElements() + 99999`. Fixed in 34b0c4ea.

File: tests/test_hgraph.cpp [hunk 10]
```diff
@@ -876,6 +877,213 @@ TEST_CASE("(PR) HGraph SearchWithRequest Reasoning", "[ft][hgraph][pr]") {
...
+    int64_t found_label = ids[0];
+
+    req.expected_labels_ = {found_label};
+    auto result = index->SearchWithRequest(req);
+    REQUIRE(result.has_value());
+    REQUIRE_FALSE(result.value()->GetReasoning().empty());
+
+    auto reasoning = result.value()->GetReasoning();
+    REQUIRE(reasoning.find("1/1") != std::string::npos);
+    REQUIRE(reasoning.find("0 missed") != std::string::npos);
```
[the remark this answers is not in the document]
Comment (LHT129) [reply tests/test_hgraph.cpp:915]: The reasoning report is JSON with a human-readable summary string. The current API only exposes the report as a raw JSON string without structured accessors, so string matching is the available contract. If structured accessors are added later, these tests can migrate.

File: src/impl/reasoning/search_reasoning_test.cpp [hunk 11]
```diff
@@ -271,4 +271,164 @@ TEST_CASE("ReasoningContext termination reasons are centralized", "[reasoning]")
...
+    SECTION("Zero distance sentinel: second visit overwrites zero first distance") {
+        // Known limitation: RecordVisit uses 0.0F as sentinel for "not yet set",
+        // so a true zero distance gets overwritten by the next visit.
+        ctx.RecordVisit(0, 0.0F, 1);
+        ctx.RecordVisit(0, 0.5F, 2);
+
+        auto it = ctx.expected_traces_.find(0);
+        REQUIRE(it.value().quantized_distance == 0.5F);
+        REQUIRE(it.value().visited_at_hop == 2);
+    }
```
[the remark this answers is not in the document]
Comment (LHT129) [reply src/impl/reasoning/search_reasoning_test.cpp:383]: This is intentionally documenting the current behavior (not endorsing it). The test section is named "Zero distance sentinel: second visit overwrites zero first distance" with a comment explaining the known limitation. This makes the behavior visible and ensures any future fix will be caught by test changes rather than silent regression.

File: src/impl/reasoning/search_reasoning_test.cpp [hunk 12]
```diff
@@ -271,4 +271,164 @@ TEST_CASE("ReasoningContext termination reasons are centralized", "[reasoning]")
...
 }
 
+TEST_CASE("ReasoningContext SetTermination", "[reasoning]") {
+    DefaultAllocator allocator;
+    ReasoningContext ctx(&allocator);
+
+    SECTION("SetTermination stores reason") {
+        REQUIRE(ctx.termination_reason_.empty());
+        ctx.SetTermination(ReasoningContext::kTerminationLowerBoundReached);
+        REQUIRE(ctx.termination_reason_ == "lower_bound_reached");
```
[the remark this answers is not in the document]
Comment (LHT129) [reply src/impl/reasoning/search_reasoning_test.cpp:281]: These are intentionally white-box unit tests for ReasoningContext internals. The class exposes its members as public for this purpose. The functional tests in test_hgraph.cpp exercise the public API (GenerateReport / GetReasoning).

**Review by wxyucs — APPROVED** [review]
lgtm
